src/adr/0018-jmap-new-specs.md

```diff
@@ -0,0 +1,48 @@
+# 18. New JMAP specifications adoption
+
+Date: 2020-02-06
+
+## Status
+
+Proposed
+
+## Context
+
+Historically, James is an early adopter for the JMAP specification, and a first partial implementation was conducted when JMAP was just a draft. But with time, the IETF draft went with radical changes and the community could not keep this implementation up to date with the spec changes.
+
+As off summer 2019, JMAP core ([RFC 8620](https://tools.ietf.org/html/rfc8620)) and JMAP mail ([RFC 8621](https://tools.ietf.org/html/rfc8621)) have been officially published. Thus we should implement these new specifications.
+
+We need to keep in mind though that part of the community actively relies on the actual 'draft' implementation of JMAP existing in James. We should ensure no changes is done to that 'draft' protocol is done while implementing the new one.
+
+## Decision
+
+The proposed approach would be as follow:
+
+* Rename packages `server/protocols/jmap*` and guice packages `server/container/guice/protocols/jmap*` to `jmap-draft`. `JMAPServer` should also be renamed to `JMAPDraftServer` (this has already been contributed [here](https://github.com/apache/james-project/pull/164), thanks to @cketti). No modification in `jmap-draft` integration tests should be conducted afterwards.
+* Create a new `jmap` package, binding it on a new port.
+* Implement the new JMAP request structure with the echo method
+* Implement authentication and session of the new JMAP protocol
+* Implement protocol-level error handling
+* Port existing mailbox methods of `jmap-draft` to `jmap`
+* Port existing email methods of `jmap-draft` to `jmap`
+* Port existing vacation methods of `jmap-draft` to `jmap`
+* Support uploads/downloads
+
+Then when we finish to port our existing methods to the new JMAP specifications, we can implement the new features that we are missing and are absolutely needed:
+
+* EmailSubmission
+* Push and queryChanges
+* Threads
+
+In order to ease testing and development speed, we decided to support `jmap` only on top of memory-guice at first. Then, we would add it for distributed-james Guice product, and other products might follow later.
+
```

Comment:
Now that James is using reacting programming more and more, should we write the new `jmap` implementation using Reactor? Would it help for Push implementation?
Are we using Java or Scala?
Do we replace jetty with a reactive http server?

Reply:
That's legitimate questions, and we should have a discussion about it I guess. 

* Because we tend to push james more towards Reactor, it would make sense to use it for jmap
* Would it help push? I honestly don't know, as we don't have an existing implementation of push to compare to and I'm still not comfortable enough with Reactor in general
* Scala could be cool I guess, but the Robustas has not much knowledge about it compared to the Mokas. It could be a good opportunity to strengthen our skills on it, but will slow down the dev as well. So how much time and resources are we willing to spend on that too?
* No idea on your last point (jetty vs reactive http server) TBH

Reply:
JMAP relies on defining many objects. Even if the first step in Scala could be harder, the gain using Scala for defining objects would maybe compensate for it.

Reply:
> That's legitimate questions, and we should have a discussion about it indeed.
> 
>     * Because we tend to push james more towards Reactor, it would make sense to use it for jmap

It will help a lot with performance in the middle term (when we'll get rid of `.block` calls)
 
>     * Would it help push? I honestly don't know, as we don't have an existing implementation of push to compare to and I'm still not comfortable enough with Reactor in general

Push with Reactor would be publishing a Flux.

>     * Scala could be cool I guess, but the Robustas has not much knowledge about it compared to the Mokas. It could be a good opportunity to strengthen our skills on it, but will slow down the dev as well. So how much time and resources are we willing to spend on that too?

I'm always eager to share knowledge. The question is more about you.

> 
>     * No idea on your last point (jetty vs reactive http server) TBH

I think it doesn't make much sense to keep using jetty

Reply:
+1 for scala

Reactive server adoption would make harder serving  both version on the same endpoints IMO

If we implement it, we will target distributed-james + memory for integration testing (not cassandra-guice). If another contributor want to contribute to this effort, he can also use persistant memory data-structure. It should be a OR.

Reply:
> I'm always eager to share knowledge. The question is more about you.

No problem about that, happy to do some Scala for a change :)

Reply:
> Reactive server adoption would make harder serving both version on the same endpoints IMO

True. That's an argument for two-ports solution

Reply:
Two ports might mean harder adoption on frontend side?

Reply:
FYI we decided to go forward with Scala as a team

Reply:
Regarding two ports vs one port, I'm confident that porting existing jmap code to another server technology is not very costly, the split is quite well done regarding business/technical code.

Reply:
So the proposal could be:

1. Port jmap-draft to be served with a reactive technology
2. Implement a JMAP meta project to select the jmap version specified in the accept header and 'map' it to the correct implementation
3. Start crafting the 'new' JMAP spec, isolated in it's maven package.